src/network/priorities.rs

```diff
@@ -0,0 +1,79 @@
+use std::{
+    cmp::{max, min},
+    collections::HashMap,
+    sync::Arc,
+};
+
+use contract_client::PeerId;
+use static_assertions::const_assert;
+
+use crate::cli::Config;
+
+type Priority = i8;
+
+/// The number subtracted from the priority while the worker is busy executing a query
+const LEASE_PENALTY: Priority = 2;
+
+pub struct WorkersPool {
+    config: Arc<Config>,
+    priorities: HashMap<PeerId, Priority>,
+}
+
+impl WorkersPool {
```

Comment:
My concern regarding the chosen method of evaluation of worker performance is that it is hard to reason about impact of the distant past behaviour.

For example, a worker could have a hiccup, but yet to receive quite a bit of queries at that moment (e.g. because it hosts some low replicated data chunks, that got a sudden demand). In such case, the worker will be significantly de-prioritised, but for how long? Might be for a very long time, depending on a usage pattern and overall network performance...

I suggest to track workers with a set of windowed (by time) counters and to compute resulting priorities at query time with some formula.

A set of counters could be:

* Number of "big" timeouts
* Number of times the worker failed to respond within a retry threshold
* Number of times worker responded within a 50% percentile
* Number of Busy errors
* Number of currently pending requests made to to this worker

Reply:
These priorities are reset every epoch, which is 30mins currently, but I suspect it may get shorter later. This simple algo has already sped up the streaming about 2-4 times. 

I like what you're suggesting, but how about getting back to it after the initial release? It looks like improving performance without ability to run real world benchmarks

Reply:
Done